HW_2.1/hw_1.py

```diff
@@ -0,0 +1,393 @@
+
+Напишите класс `AminoAcidSequence`:
++ Данный класс реализует интерфейс `BiologicalSequence`
++ Добавьте этому классу один любой метод, подходящий по смыслу к аминокислотной последовательности. Например, метод для нахождения изоэлектрической точки, молекулярного веса и т.д.
+
+Комментарий по поводу метода `NucleicAcidSequence.complement`, так как я хочу, чтобы вы сделали его опредедённым образом:
+
+При вызове `dna.complement()` или условного `dna.check_alphabet()` должны будут вызываться соответствующие методы из `NucleicAcidSequence`. При этом, данный метод должен обладать свойством полиморфизма, иначе говоря, внутри `complement` не надо делать условия а-ля `if seuqence_type == "DNA": return self.complement_dna()`, это крайне не гибко. Данный метод должен опираться на какой-то общий интерфейс между ДНК и РНК. Создание экземпляров `NucleicAcidSequence` не подразумевается, поэтому код `NucleicAcidSequence("ATGC").complement()` не обязан работать, а в идеале должен кидать исключение `NotImplementedError` при вызове от экземпляра `NucleicAcidSequence`
+
+Вся сложность задания в том, чтобы правильно организовать код. Если у вас есть повторяющийся код в сестринских классах или родительском и дочернем, значит вы что-то делаете не так.
+
+
+Маленькое замечание: По-хорошему, между классом `BiologicalSequence` и классами `NucleicAcidSequence` и `AminoAcidSequence`, ещё должен быть класс-прослойка, частично реализующий интерфейс `BiologicalSequence`, но его писать не обязательно, так как задание и так довольно большое (правда из-за этого у вас неминуемо возникнет повторяющийся код в классах `NucleicAcidSequence` и `AminoAcidSequence`)
+"""
+
+from abc import ABC, abstractmethod
+
+class NotInAlphabetException(Exception):
+    pass
+
+class BiologicalSequence(ABC):
+
+    def __init__(self, seq):
+        self.seq = seq.lower()
+
+    def __len__(self):
+        return len(self.seq)
+
+    def __getitem__(self, item):
+        return self.seq.__getitem__(item)
+
+    def __str__(self):
+        return self.seq
+
+    def print(self):
+        print(self.seq)
+
+    @abstractmethod
+    def check_alphabet(self):
+        pass
+
+
+class AlphabetAwareSequence(BiologicalSequence, ABC):
+
+    def __init__(self, seq, alphabet):
+        super(AlphabetAwareSequence, self).__init__(seq)
+        self.alphabet = alphabet
+
+    def check_alphabet(self):
+        for i in self.seq:
+            if i not in self.alphabet:
+                return False
+        return True
+
+
+class NucleicAcidSequence(AlphabetAwareSequence, ABC):
+
+    def __init__(self, seq, alphabet, complements):
+        super(NucleicAcidSequence, self).__init__(seq, alphabet)
+        self.alphabet = alphabet
+        self.complements = complements
+
+    def complement(self):
+        try:
+            return NucleicAcidSequence("".join(map(lambda i: self.complements[i], self.seq)), self.alphabet, self.complements)
```

Comment:
Если сделать вот так, то будет возвращаться объект того же типа, от которого был вызван метод. Текущая реализация не очень логична, так как тип потенциально меняется 
```suggestion
            return type(self)("".join(map(lambda i: self.complements[i], self.seq)), self.alphabet, self.complements)
```